src/translate/axiom_rules.py

```diff
@@ -1,29 +1,315 @@
+import options
 import pddl
 import sccs
 import timers
 
 from collections import defaultdict
+from itertools import chain
 
 
 DEBUG = False
 
+class AxiomDependencies(object):
+    def __init__(self, axioms):
+        if DEBUG:
+            assert all(isinstance(axiom.effect, pddl.Atom) for axiom in axioms)
+        self.derived_variables = {axiom.effect for axiom in axioms}
+        self.positive_dependencies = defaultdict(set)
+        self.negative_dependencies = defaultdict(set)
+        for axiom in axioms:
+            head = axiom.effect
+            for body_literal in axiom.condition:
+                body_atom = body_literal.positive()
+                if body_atom in self.derived_variables:
+                    if body_literal.negated:
+                        self.negative_dependencies[head].add(body_atom)
+                    else:
+                        self.positive_dependencies[head].add(body_atom)
+
+    # Remove all information for variables whose literals are not necessary.
+    # We do not need to remove single entries from the dicts because if the key
+    # (= head of an axiom) is relevant, then all its values (= body of axiom)
+    # must be relevant by definition.
+    def remove_unnecessary_variables(self, necessary_literals):
+        for var in self.derived_variables.copy():
+            if var not in necessary_literals and var.negate() not in necessary_literals:
+                self.derived_variables.remove(var)
+                self.positive_dependencies.pop(var, None)
+                self.negative_dependencies.pop(var, None)
+
+
+class AxiomCluster(object):
+    def __init__(self, derived_variables):
+        self.variables = derived_variables
+        self.axioms = dict((v, []) for v in derived_variables)
+        # Positive children will be populated with clusters that contain an
+        # atom that occurs in the body of an axiom whose head is from this
+        # cluster. Negative children analogous for atoms that occur negated
+        # in the body.
+        self.positive_children = set()
+        self.negative_children = set()
+        self.needed_negatively = False
+        self.layer = 0
+
+
+def handle_axioms(operators, axioms, goals, layer_strategy):
+    clusters = compute_clusters(axioms, goals, operators)
+    axiom_layers = compute_axiom_layers(clusters, layer_strategy)
+
+    # TODO: It would be cleaner if these negated rules were an implementation
+    # detail of the heuristics in the search component that make use of them
+    # rather than part of the translation process. They should be removed in
+    # the future. Similarly, it would be a good idea to remove the notion of
+    # axiom layers and derived variable default values from the output.
+    # (All derived variables should be binary and default to false.)
```

Comment:
It would be good to create an issue -- if we don't have one yet -- for this potential follow-up work (removing negated rules from the translator, removing axiom layers from the output, removing default values for axioms from the output, making all derived variables binary). I think Augusto expressed an interest on working on translator issues, and I think this would be a nice improvement, even though it's mostly under the hood.